Add persistence of collapsed state of space panel

gefgu opened the pull request:
Signed-off-by: gefgu <gefgu@hotmail.com>

<!-- Thanks for submitting a PR! Please ensure the following requirements are met in order for us to review your PR -->

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [X] Linter and other CI checks pass
* [X] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

<!--
If you would like to specify text for the changelog entry other than your PR title, add the following:

Notes: Add super cool feature

Changes in this project also generate changelogs in Element Web. To disable this, use the following:

element-web notes: none

or specify alternative text:

element-web notes: Add super cool feature
-->

Notes: Add persistence of collapsed state of space panel. It fixes vector-im/element-web#23172

Type: enhancement

Behavior: (Hard reload was made by `ctrl + shift + r`)

https://user-images.githubusercontent.com/53129852/188504813-b9f60b83-52fa-484c-ae26-2d864db1813b.mp4

This change was made using LocalStorage implemented inside of the component. I'm not sure it's the best solution because of most of the other LocalStorage values are implemented using stores. In this way, I want feedback about my approach and I'm willing to keep working on this issue until it is resolved.

<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * Add persistence of collapsed state of space panel. It fixes vector-im/element-web#23172 ([\#9245](https://github.com/matrix-org/matrix-react-sdk/pull/9245)). Contributed by @gefgu.<!-- CHANGELOG_PREVIEW_END -->

turt2live commented on src/components/views/spaces/SpacePanel.tsx:
```diff
@@ -337,6 +349,14 @@ const SpacePanel = () => {
     });
 
+    useEffect(() => {
+        try {
+            localStorage.setItem(KEY_PANEL_IS_COLLAPSED, isPanelCollapsed);
+        } catch (error) {
+            console.log(error);
+        }
```
what are we expecting to fail with these localstorage calls?

gefgu replied:
I read through some blog posts and this catch behavior was common. So, I implemented it. Now, that you asked about, it doesn't seems necessary. I'm going to write the cypress test and if everything is ok, I think I will remove this try/catch block. What do you think?

turt2live replied:
That seems like a good way forward, yea. There's a chance that localstorage will break, but also we don't guard it elsewhere.

Not that it really helps the safety here, we could also move the localstorage key over to the Settings system as a device-only setting, making it safer in the future (if we ever add safety).

turt2live reviewed (changes requested):
Thanks for the contribution! It looks pretty straightforward, though I think a simple cypress test or similar would be a great addition to this.

gefgu commented:
I've added the cypress test and would be delighted if you could review it - It is the first time I work with cypress :). Also, I've removed the try/catch block. It does not seems to be necessary.

turt2live commented on cypress/e2e/space-panel/space-panel.spec.ts:
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
```
This should be your copyright, not ours :)

gefgu replied:
I've copied this from the other test files because all I've checked had this license above them. This is my first time adding a copyright in a PR, should I add my name in the place of "The Matrix.org Foundation C.I.C."?

turt2live replied:
Yup, with your email address if you prefer:

```
Copyright 2022 Your Name
```
or
```
Copyright 2022 Your Name <email@example.org>
```

turt2live commented on src/components/views/spaces/SpacePanel.tsx:
```diff
@@ -337,6 +344,14 @@ const SpacePanel = () => {
     });
 
+    useEffect(() => {
+        try {
+            localStorage.setItem(KEY_PANEL_IS_COLLAPSED, isPanelCollapsed);
+        } catch (error) {
+            console.log(error);
+        }
```
looks like the try/catch is still here on the setter

turt2live reviewed (commented):
generally this looks good to me - thanks! 

I've merged develop in hopes of kicking the CI into working properly - apologies if this interferes with your workflow.

turt2live commented on cypress/e2e/space-panel/space-panel.spec.ts:
```diff
@@ -0,0 +1,43 @@
+        cy.get(".mx_SpacePanel_toggleCollapse").click();
+        cy.get(".mx_SpacePanel").should("not.have.class", "collapsed");
+        cy.reload(true);
+        cy.get(".mx_SpacePanel").should("not.have.class", "collapsed");
```
we should also probably check to see if clicking it again and reloading causes it to un-collapse

turt2live reviewed (commented):
still looks good to me overall - the CI is very unhappy for some reason though.
